drivers/regulator/regulator_rpi_pico.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_core_supply_regulator
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/dt-bindings/regulator/rpi_pico.h>
+#include <zephyr/sys/linear_range.h>
+#include <zephyr/toolchain.h>
+#include <hardware/regs/vreg_and_chip_reset.h>
+#include <hardware/structs/vreg_and_chip_reset.h>
+
+#define VREG_VSEL_IDX_OFFSET 5
+
+static const struct linear_range core_ranges[] = {
+	LINEAR_RANGE_INIT(800000u, 50000u, 0u, 10u),
+};
+
+struct regulator_rpi_pico_config {
+	struct regulator_common_config common;
+	const struct linear_range *ranges;
+	const size_t num_ranges;
+	const uint32_t reg;
+	const bool brown_out_detection;
+	const uint32_t brown_out_threshold;
+};
+
+struct regulator_rpi_pico_data {
+	struct regulator_common_data data;
+};
+
+/*
+ * APIs
+ */
+
+static unsigned int regulator_rpi_pico_count_voltages(const struct device *dev)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+
+	return linear_range_group_values_count(config->ranges, config->num_ranges);
+}
+
+static int regulator_rpi_pico_list_voltage(const struct device *dev, unsigned int idx,
+					   int32_t *volt_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+
+	return linear_range_group_get_value(config->ranges, config->num_ranges, idx, volt_uv);
+}
+
+static int regulator_rpi_pico_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+	volatile uint32_t *vreg = &(((volatile vreg_and_chip_reset_hw_t *)config->reg)->vreg);
+	uint16_t idx;
+	int ret;
+
+	ret = linear_range_group_get_win_index(config->ranges, config->num_ranges, min_uv, max_uv,
+					       &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*vreg = ((*vreg & ~VREG_AND_CHIP_RESET_VREG_VSEL_BITS) |
+		 ((idx + VREG_VSEL_IDX_OFFSET) << VREG_AND_CHIP_RESET_VREG_VSEL_LSB));
+
+	return 0;
+}
+
+static int regulator_rpi_pico_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+	volatile uint32_t *vreg = &(((volatile vreg_and_chip_reset_hw_t *)config->reg)->vreg);
+
+	return linear_range_group_get_value(
+		config->ranges, config->num_ranges,
+		((*vreg & VREG_AND_CHIP_RESET_VREG_VSEL_BITS) >> VREG_AND_CHIP_RESET_VREG_VSEL_LSB),
```

Comment:
I don't have a board to test on, but won't you need to subtract `VREG_VSEL_IDX_OFFSET` here to get the correct voltage? As an aside, the linear range API also supports multiple values mapping to the same voltage (see here for an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/regulator/regulator_pca9420.c#L127), which might be the best way to handle this regulator

Reply:
Thanks for suggestion.
I rewrite with correctly linear range API usage, and removed `VREG_VSEL_IDX_OFFSET`.